tools/istio-iptables/pkg/capture/run.go

```diff
@@ -754,38 +756,215 @@ func (cfg *IptablesConfigurator) executeIptablesCommands(iptVer *dep.IptablesVer
+		}
+		for _, el := range flagsAndValues {
+			if el.flag == "A" || el.flag == "-append" || el.flag == "I" || el.flag == "-insert" {
+				chain = el.value
+				break
+			}
+		}
+		if chain == "" {
+			continue
+		}
+
+		// Construct the sorted rule
+		var sortedRule string
+		sort.Slice(flagsAndValues, func(i, j int) bool {
+			return flagsAndValues[i].flag < flagsAndValues[j].flag
+		})
+		for _, fv := range flagsAndValues {
+			sortedRule += fmt.Sprintf("-%s %s ", fv.flag, fv.value)
+		}
+		sortedRule = strings.TrimSpace(sortedRule)
+
+		_, ok := result[table][chain]
+		if !ok {
+			result[table][chain] = []string{}
+		}
+		result[table][chain] = append(result[table][chain], sortedRule)
+
+	}
+	return result
+}
+
+func (cfg *IptablesConfigurator) VerifyRerunStatus(iptVer, ipt6Ver *dep.IptablesVersion) (bool, bool) {
+	applyRequired := false
+	residueFound := false
+
+check_loop:
+	for _, ipCfg := range []struct {
+		ver        *dep.IptablesVersion
+		expected   string
+		checkRules [][]string
+	}{
+		{iptVer, cfg.ruleBuilder.BuildV4Restore(), cfg.ruleBuilder.BuildCheckV4()},
+		{ipt6Ver, cfg.ruleBuilder.BuildV6Restore(), cfg.ruleBuilder.BuildCheckV6()},
+	} {
+		output, err := cfg.ext.RunWithOutput(constants.IPTablesSave, ipCfg.ver, nil)
+		if err == nil {
+			currentState := cfg.getStateFromSave(output.String())
+			for _, value := range currentState {
+				if residueFound {
+					break
+				}
+				residueFound = len(value) != 0
+			}
+			if !residueFound {
+				continue
+			}
+			expectedState := cfg.getStateFromSave(ipCfg.expected)
+			for table, chains := range expectedState {
+				_, ok := currentState[table]
+				if !ok {
+					applyRequired = true
+					break check_loop
+				}
+				for chain, rules := range chains {
+					currentRules, ok := currentState[table][chain]
+					if !ok || (strings.HasPrefix(chain, "ISTIO_") && len(rules) != len(currentRules)) {
+						applyRequired = true
+						break check_loop
+					}
+				}
+			}
+			err = cfg.executeIptablesCommands(ipCfg.ver, ipCfg.checkRules)
+			if err != nil {
+				applyRequired = true
+				break
+			}
+		}
+
+	}
+
+	// If there are no residue, apply step is always needed
+	if !residueFound {
+		return false, true
+	}
+
+	if residueFound && !applyRequired {
+		log.Info("Found compatible iptables rules/chains, no additional changes are needed")
+	} else if residueFound {
+		log.Info("Found residue of old iptables rules/chains, cleanup is needed")
+	}
+
+	return residueFound, applyRequired
+}
+
+func (cfg *IptablesConfigurator) executeCommands(iptVer, ipt6Ver *dep.IptablesVersion) error {
+	guardrails := false
+	defer func() {
+		if guardrails {
+			log.Info("Removing guardrails")
+			guardrailsCleanup := cfg.ruleBuilder.BuildCleanupGuardrails()
+			_ = cfg.executeIptablesCommands(iptVer, guardrailsCleanup)
+			_ = cfg.executeIptablesCommands(ipt6Ver, guardrailsCleanup)
+		}
+	}()
+
+	residueFound, applyRequired := cfg.VerifyRerunStatus(iptVer, ipt6Ver)
+	// Cleanup Step
```

Comment:
We need to explicitly error if residue == true && reconcile == false.

The three modes should be

1. Remove all without recreate
2. Apply without reconcile (error if delta - this is basically current state)
3. Apply with reconcile (warn if delta)

To represent those states we need 2 flags - `removeonly = true|false` and `reconcile = true|false`

Reply:
Done,
We now have all three as follow:
- Simply remove all without (re)create: --cleanup-only flag
- Apply all without reconcile (error if delta is found): --no-reconcile. This will: (1) "fast-fail" if delta is found and reconciliation is needed, (2) it will "fast-exit" if delta is found and no reconcilation is needed, (3) apply rules as normal if no delta is found.
- Apply all with reconcile (implies cleanup - warn if delta, then fix): default behavior, cleanup is only performed if incompatible delta is found.

btw I refactored the code to use the deltaExists, reconcile variables instead of the old leftoverFound, applyRequired